docs/guide/api-plugin.md

```diff
@@ -159,7 +159,9 @@ The following hooks are called on each incoming module request:
 - [`load`](https://rollupjs.org/plugin-development/#load)
 - [`transform`](https://rollupjs.org/plugin-development/#transform)
 
-They also have an extended `options` parameter with additional Vite-specific properties. You can read more in the [SSR documentation](/guide/ssr#ssr-specific-plugin-logic).
+These hooks also have an extended `options` parameter with additional Vite-specific properties. You can read more in the [SSR documentation](/guide/ssr#ssr-specific-plugin-logic).
+
+During dev, `resolveId` will be called with the imported `source` and the proper `importer` value when resolving during the import analysis phase. Each imported URL will be resolved while transforming a JS source and normalized to a URL the browser can handle (relative paths will be transformed to absolute paths starting from root, and out-of-root paths will be prefixed with a `/@fs/` marker). When the browser request these modules, the Vite dev server no longer have information about the original `importer`. When `resolveId` is called for these URLs, the `importer` will be an absolute path for a generic `index.html` at root.
```

Comment:
Maybe we should show the quirk first and then explain why so, so it's easier to follow the reasoning and context.

```suggestion
For ids imported by HTML files in dev, the `resolveId` hook's `importer` value will be an absolute path for a generic `index.html` at root as it's not possible to derive the actual importer due to Vite's unbundled dev server pattern. For all other ids, Vite is able to track the `importer` during the import analysis phase.
```

I'm not sure how to incorporate the other explanations as it goes very deep into Vite's internals.

Reply:
The importer is `index.html` not only for ids imported by HTML files. Although after 4.3, the new URLToModule cache in the ModuleGraph will make your proposed explanation true most of the time. But I don't think we should keep this as an optimization.

See `resolveId(id, importer)` calls for:

- [4.3.3](https://stackblitz.com/edit/vitejs-vite-qhscnr)
- [4.2.2](https://stackblitz.com/edit/vitejs-vite-xvofqc)

In 4.2.2, we don't have the cache so we end up resolving all the URLs we transpiled during import analysis again, and the importer for all these resolveId calls is `/index.html`.
```
/src/counter.ts
  \__ /home/projects/vitejs-vite-xvofqc/index.html

/src/a.ts
  \__ /home/projects/vitejs-vite-xvofqc/index.html
```

Reply:
I see. I also think it make sense to not mention it's for HTML files only too just in case.

I still think we should still re-arrange the paragraph to show what the user might encounter, then explain it though. Maybe:
```suggestion
Some `resolveId` calls' `importer` value may be an absolute path for a generic `index.html` at root as it's not always possible to derive the actual importer due to Vite's unbundled dev server pattern. For imports handled within Vite's resolve pipeline, the importer can be tracked during the import analysis phase, providing the correct `importer` value.
```

I'm not sure if this is more accurate, but trying to provide something to start off 😅